src/main/java/org/openrewrite/java/testing/junit5/AddMissingTestBeforeAfterAnnotations.java

```diff
@@ -0,0 +1,117 @@
+            @Override
+            public J.ClassDeclaration visitClassDeclaration(J.ClassDeclaration classDecl, ExecutionContext ctx) {
+                if (classDecl.getExtends() != null) {
+                    // Only classes that extend other classes can have override methods with missing annotations
+                    return SearchResult.found(classDecl);
+                }
+                return super.visitClassDeclaration(classDecl, ctx);
+            }
+        }, new AddMissingTestBeforeAfterAnnotationsVisitor());
+    }
+
+    private static class AddMissingTestBeforeAfterAnnotationsVisitor extends JavaIsoVisitor<ExecutionContext> {
+        @Override
+        public J.MethodDeclaration visitMethodDeclaration(J.MethodDeclaration method, ExecutionContext ctx) {
+            if (!method.hasModifier(J.Modifier.Type.Static) && !method.isConstructor()) {
+                Optional<Method> superMethod = TypeUtils.findOverriddenMethod(method.getMethodType());
+                if (superMethod.isPresent()) {
+                    method = maybeAddMissingAnnotation(method, superMethod.get(), LifecyleAnnotation.BEFORE_EACH, ctx);
+                    method = maybeAddMissingAnnotation(method, superMethod.get(), LifecyleAnnotation.AFTER_EACH, ctx);
+                    method = maybeAddMissingAnnotation(method, superMethod.get(), LifecyleAnnotation.TEST, ctx);
+                }
+            }
+            return super.visitMethodDeclaration(method, ctx);
+        }
+
+        private J.MethodDeclaration maybeAddMissingAnnotation(J.MethodDeclaration method, Method superMethod, LifecyleAnnotation la, ExecutionContext ctx) {
+            if (la.hasOldAnnotation(superMethod) && !la.hasNewAnnotation(method)) {
+                maybeAddImport(la.annotation);
+                return JavaTemplate.builder(la.simpleAnnotation)
+                        .javaParser(JavaParser.fromJavaVersion().classpathFromResources(ctx, "junit-jupiter-api-5.9"))
+                        .imports(la.annotation)
+                        .build()
+                        .apply(getCursor(), method.getCoordinates().addAnnotation(Comparator.comparing(J.Annotation::getSimpleName)));
+            }
+            return method;
+        }
+    }
+
+    enum LifecyleAnnotation {
+        BEFORE_EACH("org.junit.jupiter.api.BeforeEach", "org.junit.Before"),
+        AFTER_EACH("org.junit.jupiter.api.AfterEach", "org.junit.After"),
+        TEST("org.junit.jupiter.api.Test", "org.junit.Test");
+
+        String annotation;
+        String simpleAnnotation;
+        private Predicate<FullyQualified> oldAnnotationPredicate;
+        private AnnotationMatcher annotationMatcher;
+
+        LifecyleAnnotation(String annotation, String oldAnnotation) {
+            this.annotation = annotation;
+            this.simpleAnnotation = "@" + annotation.substring(annotation.lastIndexOf(".") + 1);
+            this.oldAnnotationPredicate = n -> TypeUtils.isOfClassType(n, oldAnnotation);

```

Comment:
I wonder if we should expand this predicate to also check for super class methods annotated with the new annotations. One reason is because of the ordering issue pointed out on `junit5.yml`, and another reason is because folks might have migrated by hand, and missed adding those annotations. Those overridden tests would then today not run, whereas they very likely should be.

I don't suppose there's many folks that In JUnit Jupiter intentionally override a method, and do not add the test annotation to stop running a test, although it's certainly possible. Such cases would be negatively affected there, so something to keep in mind. Ideally those folks override the super method and add `@Disabled` instead of relying on the absence of a test annotation.

Reply:
To be honest, I was first checking for the new annotations because I thought that UpdateBeforeAfterAnnotations runs first and then we just have to handle the new annotations - however this seems not to be the case, I still see the unchanged annotations.
I then looked at the other recipes in the package and had the impression, that they all look for the old stuff and migrate it to the new, so I did it also this way.
So I think all other recipes will just work if the old JUnit stuff is still in place and AddMissingTestBeforeAfterAnnotations behaves the same way.